### PR TITLE
nixos/postfix: systemd sandbox

### PR DESCRIPTION
**Pull request by tnias**
###### Motivation for this change

Add systemd service sandbox features.

###### Things done

- [x] tested on a mailserver based on [simple-nixos-mailserver](https://gitlab.com/simple-nixos-mailserver/nixos-mailserver/)

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

cc: postfix maintainers @rickynils @globin

### CODE REVIEW

File: nixos/modules/services/mail/postfix.nix
Comment (hmenke) [inline nixos/modules/services/mail/postfix.nix:56]: The sandbox is incomplete here.  This will still run `postfix` as root, although I don't know whether there is a way around it.

Comment (tnias) [reply nixos/modules/services/mail/postfix.nix:56]: I assumed it is acceptable to have no sandbox for the prepare script, because it does not run as a network daemon.
Should put it in a sandbox as well?

To my knowledge the "main daemon" has to be run as root. It drops privileges for it's children, but there is no tight systemd integration. I am open for suggestions. :)

File: nixos/modules/services/mail/postfix.nix
Comment (alyssais) [inline nixos/modules/services/mail/postfix.nix:778]: Won’t this break .forwarding, or delivering mail to a local user’s home directory?

File: nixos/modules/services/mail/postfix.nix
Comment (hmenke) [reply nixos/modules/services/mail/postfix.nix:56]: I see.  In that case I'm not sure whether there is any benefit from running this in the sandbox.

**Comment by hmenke** [comment]
It's probably a good idea to also discuss this with the people from [simple-nixos-mailserver](https://gitlab.com/simple-nixos-mailserver/nixos-mailserver) @r-raymond 
Same for your opendkim PR #93314, rspamd PR #93293, and redis PR #93277

**Comment by r-raymond** [comment]
@hmenke thanks for the heads up, I'll ping the current maintainers.

**Comment by nlewo** [comment]
Thanks for the heads up, and thanks for these patches! I really like this idea;)

Did you run the nixos-mailserver tests with your patches? I will do it this WE.

**Comment by tnias** [comment]
> Did you run the nixos-mailserver tests with your patches?

I have not yet run any of the official tests. Just tested my usecase for now.

File: nixos/modules/services/mail/postfix.nix
Comment (tnias) [reply nixos/modules/services/mail/postfix.nix:778]: Yes. Replacing `ProtectHome=true` by `ReadOnlyPaths=/home` would allow to read .forward. For delivery to the home directory this has to be disabled entirely.

Should this be behind an option? If yes, how whould an option look like?

Comment (aanderse) [reply nixos/modules/services/mail/postfix.nix:778]: I would suggest you don't put this behind an option - don't include it at all. It isn't a reasonable default for a mail server. Add documentation so users who can and wish to will know how to add it.

File: nixos/modules/services/mail/postfix.nix
Comment (aanderse) [inline nixos/modules/services/mail/postfix.nix:798]: Can't you run this in the original service as an `ExecStartPre` with the `+` prefix?

Comment (tnias) [reply nixos/modules/services/mail/postfix.nix:798]: Thanks, the `+` prefix looks good. I will fix my patches.

**Review by peti — COMMENTED** [review]
How is this kind of sandboxing going to interact with non-trivial Postfix installations that run, say a mailing list server or reflector scripts or any other mail service that needs resources other than Postfix itself?

Running the daemon in a sandbox feels like a rather intrusive change. This should definitely be opt-in.

**Comment by tnias** [comment]
> This should definitely be opt-in.

I will try to do something like [services.nginx.enableSandbox](https://github.com/NixOS/nixpkgs/blob/aed85b727980656cdccb48198a9bc847fab7ce32/nixos/modules/services/web-servers/nginx/default.nix#L466).

**Comment by aanderse** [comment]
> I will try to do something like [services.nginx.enableSandbox](https://github.com/NixOS/nixpkgs/blob/aed85b727980656cdccb48198a9bc847fab7ce32/nixos/modules/services/web-servers/nginx/default.nix#L466).

@Mic92 given `enableSandbox` was meant as a temporary measure and enough time has passed with no negative reports can we please remove it now?

@tnias I'm not in favor of module randomly adding `enableSandbox` with no consistency at all. Please see my thoughts on the matter [here](https://github.com/NixOS/nixpkgs/pull/92106#discussion_r449915484). I think @flokli articulated the issue very well with this comment:
> Sandboxing is not a boolean, but something very fuzzy.

**Comment by tnias** [comment]
The fixup changes loosen up rules to allow more general actions.

@peti:
- mailing lists (mailman3) deliver mail by lmtp (tcp or unix socket). This should not be blocked by the current rules. Are other mailinglists using something not allowed yet?
- are "reflector scripts" just forwarding mailservers? If yes, I don't see why this should not work.
- "any other mail service that needs resources other than Postfix itself" can you provide examples? So that I can loosen up the ruleset accordingly

**Comment by Izorkin** [comment]
Rules systemd.tmpfiles can be taken from here - https://github.com/NixOS/nixpkgs/pull/76604
I running this variant on my server - it worked fine.

**Comment by tnias** [comment]
I used @Izorkin's tmpfiles patch and addded some (hopefully) reasonable sandboxing options.

**Comment by tnias** [comment]
There are some permission/ownership issues in the /var/lib/postfix directory.

**Comment by nlewo** [comment]
@tnias nixos-mailserver tests passed with this patchset.

**Comment by tnias** [comment]
@nlewo I just cleared my /var/lib/postfix directory and now it works just fine. I probably messed up the permissions at some point. ¯\_(ツ)_/¯

**Review by nlewo — APPROVED** [review]

**Comment by tnias** [comment]
I ran into the `fatal: open lock file pid/inet.submission: cannot open file: Permission denied` issue again. This widens the sandbox, so it hopefully won't happen again.

File: nixos/modules/services/mail/postfix.nix
Comment (Mic92) [inline nixos/modules/services/mail/postfix.nix:787]: Does this still allow other milters? I know that our rspamd integration for postfix for example creates a socket in `/run/rspamd/rspamd-milter.sock`, opendkim might be similar.

Comment (Mic92) [reply nixos/modules/services/mail/postfix.nix:787]: @GrahamcOfBorg test rspamd

**Comment by Mic92** [comment]
@GrahamcOfBorg test rspamd

File: nixos/modules/services/mail/postfix.nix
Comment (nlewo) [reply nixos/modules/services/mail/postfix.nix:787]: IIUC, the `full` value still allows RW accesses to `/run`. 
Also, I've run the nixos-mailserver tests (rspamd, opendkim, dovecot,...) on top of this patch and tests passed.

**Comment by Mic92** [comment]
I would also say to have it tested in unstable for a bit longer and merge it after the branch-off. Postfix itself already does quite a bit of privilege separation so these sandboxes would not improve too much in terms of security.

**Comment by ju1m** [comment]
Due to the use of `ExecStartPre=` (as a `str`) this breaks configs using it as a `str`, or using `preStart=`. The error is:
```
The option `systemd.services.postfix.serviceConfig.ExecStartPre' has conflicting definitions, in `/nix/store/pb48ks5whk0sk5nn5602m6xw307vjbaj-nixpkgs-patched/nixos/modules/system/boot/systemd.nix' and `/nix/store/pb48ks5whk0sk5nn5602m6xw307vjbaj-nixpkgs-patched/nixos/modules/services/mail/postfix.nix'.
```
Such option merging conflict happens regularly to me and others https://github.com/NixOS/nixpkgs/pull/97371
So this does not concern only this PR (though at least it may be more user friendly to set `services.postfix`'s `ExecStartPre=` as a `listOf str`). One general solution could be to modify the merging of systemd unit options to merge `str` and `listOf str` into `listOf str`. Or at least also make `preStart=` generate an `ExecStartPre=` which is a `listOf str`.

**Comment by Mic92** [comment]
Was this closed on purpose?

**Comment by tnias** [comment]
@Mic92 There seemed to be little action around it and in the spirit of reducing the number of open PRs I closed it. We can also keep it open if you prefer.

**Comment by Mic92** [comment]
@tnias I think it just take some time to make sure that postfix does not break.
/marvin opt-in
/status needs_reviewer

**Comment by aanderse** [comment]
We're roughly 4-5 months away from a new release. This change presents high enough risk that I would **love** to see it merged very soon.  4-5 months sounds like a long enough time to break and fix enough machines that we don't have to worry when this goes stable.

Just a thought...

File: nixos/modules/services/mail/postfix.nix
Comment (Mic92) [inline nixos/modules/services/mail/postfix.nix:773]: @tnias Have you checked if postfix drops these capabilities for its less-privileged childs?
Otherwise it might make postfix less secure than it was before (those capabilities allow to read/write arbitrary files and change to different users/groups).

Comment (Izorkin) [reply nixos/modules/services/mail/postfix.nix:773]: Before this update
Master process:
```
CapInh: 0000000000000000
CapPrm: 000001ffffffffff
CapEff: 000001ffffffffff
CapBnd: 000001ffffffffff
CapAmb: 0000000000000000
```

Slave process:
```
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 000001ffffffffff
CapAmb: 0000000000000000
```

```
0x000001ffffffffff=cap_chown,cap_dac_override,cap_dac_read_search,cap_fowner,cap_fsetid,cap_kill,cap_setgid,cap_setuid,cap_setpcap,cap_linux_immutable,cap_net_bind_service,cap_net_broadcast,cap_net_admin,cap_net_raw,cap_ipc_lock,cap_ipc_owner,cap_sys_module,cap_sys_rawio,cap_sys_chroot,cap_sys_ptrace,cap_sys_pacct,cap_sys_admin,cap_sys_boot,cap_sys_nice,cap_sys_resource,cap_sys_time,cap_sys_tty_config,cap_mknod,cap_lease,cap_audit_write,cap_audit_control,cap_setfcap,cap_mac_override,cap_mac_admin,cap_syslog,cap_wake_alarm,cap_block_suspend,cap_audit_read,cap_perfmon,cap_bpf,cap_checkpoint_restore
```

After this update:
Master process:
```
CapInh: 0000000000000000
CapPrm: 00000000000004c2
CapEff: 00000000000004c2
CapBnd: 00000000000004c2
CapAmb: 0000000000000000
```

Slave process:
```
CapInh: 0000000000000000
CapPrm: 0000000000000000
CapEff: 0000000000000000
CapBnd: 00000000000004c2
CapAmb: 0000000000000000
```

```
0x00000000000004c2=cap_dac_override,cap_setgid,cap_setuid,cap_net_bind_service
```

**Pull request closed without merging**
